fendermint/vm/interpreter/src/signed.rs

```diff
@@ -17,6 +20,67 @@ pub struct InvalidSignature(pub String);
 pub type SignedMessageApplyRet = Result<FvmApplyRet, InvalidSignature>;
 pub type SignedMessageCheckRet = Result<FvmCheckRet, InvalidSignature>;
 
+/// Different kinds of signed messages.
+///
+/// This technical construct was introduced so we can have a simple linear interpreter stack
+/// where everything flows through all layers, which means to pass something to the FVM we
+/// have to go through the signature check.
+pub enum VerifiableMessage {
+    /// A normal message sent by a user.
+    Signed(SignedMessage),
+    /// Something we constructed to pass on to the FVM.
+    Synthetic(SyntheticMessage),
+}
+
+impl VerifiableMessage {
+    pub fn verify(&self, chain_id: &ChainID) -> Result<(), SignedMessageError> {
+        match self {
+            Self::Signed(m) => m.verify(chain_id),
+            Self::Synthetic(m) => m.verify(chain_id),
+        }
+    }
+
+    pub fn into_message(self) -> FvmMessage {
+        match self {
+            Self::Signed(m) => m.into_message(),
+            Self::Synthetic(m) => m.message,
+        }
+    }
+}
+
+pub struct SyntheticMessage {
+    /// The artifical message.
+    message: FvmMessage,
+    /// The CID of the original message (assuming here that that's what was signed).
+    orig_cid: cid::Cid,
+    /// The signature over the original CID.
+    signature: Signature,
+}
+
+impl SyntheticMessage {
+    pub fn new<T: Serialize>(
+        message: FvmMessage,
+        orig: &T,
+        signature: Signature,
+    ) -> Result<Self, IpldError> {
+        let orig_cid = fendermint_vm_message::cid(orig)?;
+        Ok(Self {
+            message,
+            orig_cid,
+            signature,
+        })
+    }
+
+    pub fn verify(&self, chain_id: &ChainID) -> Result<(), SignedMessageError> {
+        let mut data = self.orig_cid.to_bytes();
+        data.extend(chain_id_bytes(chain_id).iter());
```

Comment:
I thought the chainID was a prefix instead of a suffix of the message cid. Out of a curiosity, is this Fendermint-specific or Filecoin also does this?

Reply:
I have no idea what Filecoin does, I did this in https://github.com/consensus-shipyard/fendermint/pull/115

Ethereum seems to play with the signature itself, which seemed so complicated I didn't think we need to replicate it in Fendermint. The `ethers` library takes care of it for us, but for regular messages I do the suffix. Is prefix safer?